File: Data Structures & Algorithms/climbing-stairs/submission-4.cpp
class Solution {
public:
    int climbStairs(int n) {
        vector<int> dp(n+1, 0);
       if(n <= 2) return n;

       dp[1] = 1;
       dp[2] = 2;

       for(int i = 3;i<=n;i++){
        dp[i] = dp[i-1]+dp[i-2];
       }

       return dp[n];
    }

    int helper(vector<int> &dp, int curr, int n) {
        if(curr == n) return 1;
        if(curr > n) return 0;
        if(dp[curr] != -1)
            return dp[curr];
        int ans = helper(dp, curr+1, n);
        ans+= helper(dp, curr+2,n);

        return dp[curr] = ans;
    }
};

File: Data Structures & Algorithms/generate-parentheses/submission-4.java
class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> ans = new ArrayList<>();
        dfs(n, "", 0,0,ans);
        return ans;
    }

    public void dfs(int n, String tmp, int open, int close, List<String> ans) {
        if(open < close || open+close >2*n || open > n) {
            return;
        }

        if(open + close == n*2 && open == close) {
            ans.add(tmp);
            return;
        }

        String tmp1 = tmp + '(';
        dfs(n , tmp1, open+1, close, ans);
        tmp1 = tmp + ')' ;
        dfs(n , tmp1, open, close+1, ans);
        return;
    }
}
